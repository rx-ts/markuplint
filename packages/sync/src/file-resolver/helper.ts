import path from 'path'

import { Config, ConfigSet, margeConfig } from '@markuplint/file-resolver'
import { cosmiconfigSync } from 'cosmiconfig'

import { tryRequirePkg } from '../helper'

import { loadConfigFile } from './load-config-file'

const explorer = cosmiconfigSync('markuplint')

export function search<T = Config>(dir: string, cacheClear: boolean) {
  if (!cacheClear) {
    explorer.clearCaches()
  }
  dir = path.dirname(dir)
  const result = explorer.search(dir)
  if (!result || result.isEmpty) {
    return null
  }
  return {
    filePath: result.filepath,
    config: result.config as T,
  }
}

export function load<T = Config>(filePath: string, cacheClear: boolean) {
  if (!cacheClear) {
    explorer.clearCaches()
  }
  const result = explorer.load(filePath)
  if (!result || result.isEmpty) {
    return null
  }
  return {
    filePath: result.filepath,
    config: result.config as T,
  }
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export function recursiveLoad(
  config: Config,
  filePath: string,
  files: Set<string>,
  cacheClear: boolean,
): ConfigSet {
  const errs: Error[] = []
  const baseDir = path.dirname(filePath)
  if (config.extends) {
    const extendFiles = Array.isArray(config.extends)
      ? config.extends
      : [config.extends]
    for (const _file of extendFiles) {
      if (/^\.+\//.test(_file)) {
        const file = path.resolve(path.join(baseDir, _file))
        if (files.has(file)) {
          continue
        }
        const extendFileResult = loadConfigFile(file, true, cacheClear)
        if (!extendFileResult) {
          continue
        }
        files = new Set(files).add(file)
        config = margeConfig(extendFileResult.config, config)
      } else {
        try {
          const mod = tryRequirePkg<Config>(_file, true)
          files.add(_file)
          config = margeConfig(mod, config)
        } catch (err) {
          errs.push(err as Error)
        }
      }
    }
  }
  delete config.extends
  return {
    files,
    config,
    errs,
  }
}
