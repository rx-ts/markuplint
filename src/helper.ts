export function tryRequirePkg<T>(pkg: string): T | undefined
export function tryRequirePkg<T>(pkg: string, throwError: true): T
export function tryRequirePkg<T>(
  pkg: string,
  throwError?: boolean,
): T | undefined {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
    const result = require(pkg) as T | { default: T }
    return 'default' in result ? result.default : result
  } catch (err) {
    if (throwError) {
      throw err
    }
  }
}
