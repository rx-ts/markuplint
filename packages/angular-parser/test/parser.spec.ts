import { cleanParse } from './helper'

describe('parser', () => {
  test('parse', () => {
    expect(
      cleanParse(/* HTML */ `<!DOCTYPE html>
        <![CDATA[ Within this Character Data block I can use double dashes as
        much as I want (along with <, &, ', and ") *and* %MyParamEntity; will be
        expanded to the text "Has been expanded" ... however, I can't use the
        CEND sequence. If I need to use CEND I must escape one of the brackets
        or the greater-than sign using concatenated CDATA sections. ]]>
        <!-- top level comment -->
        <html lang="en">
          <body>
            <!-- body comment -->
            <input [type]="'text'" />
            <div>
              Text1
              <span>Text</span>
              Text2
            </div>
          </body>
        </html>`),
    ).toMatchSnapshot()
  })

  test('invalid tag', () => {
    expect(cleanParse('<a/></>')).toMatchSnapshot()
  })
})
