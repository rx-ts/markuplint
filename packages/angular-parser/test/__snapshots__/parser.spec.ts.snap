// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser doctype 1`] = `
Object {
  "isFragment": false,
  "nodeList": Array [
    Object {
      "endCol": 103,
      "endLine": 1,
      "endOffset": 102,
      "isFragment": false,
      "isGhost": false,
      "name": "html",
      "nodeName": "#doctype",
      "publicId": "-//W3C//DTD HTML 4.01 Transitional//EN",
      "raw": "<!DOCTYPE html public \\"-//W3C//DTD HTML 4.01 Transitional//EN\\" \\"http://www.w3.org/TR/html4/loose.dtd\\">",
      "startCol": 1,
      "startLine": 1,
      "startOffset": 0,
      "systemId": "http://www.w3.org/TR/html4/loose.dtd",
      "type": "doctype",
    },
  ],
}
`;

exports[`parser doctype 2`] = `
Object {
  "isFragment": false,
  "nodeList": Array [
    Object {
      "endCol": 64,
      "endLine": 1,
      "endOffset": 63,
      "isFragment": false,
      "isGhost": false,
      "name": "html",
      "nodeName": "#doctype",
      "publicId": "-//W3C//DTD HTML 4.01 Transitional//EN",
      "raw": "<!DOCTYPE html public \\"-//W3C//DTD HTML 4.01 Transitional//EN\\">",
      "startCol": 1,
      "startLine": 1,
      "startOffset": 0,
      "systemId": "",
      "type": "doctype",
    },
  ],
}
`;

exports[`parser doctype 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -15,49 +15,7 @@
        \\"startLine\\": 1,
        \\"startOffset\\": 0,
        \\"systemId\\": \\"http://www.w3.org/TR/html4/loose.dtd\\",
        \\"type\\": \\"doctype\\",
      },
-     Object {
-       \\"endCol\\": 103,
-       \\"endLine\\": 1,
-       \\"endOffset\\": 102,
-       \\"isFragment\\": false,
-       \\"isGhost\\": true,
-       \\"namespace\\": \\"http://www.w3.org/1999/xhtml\\",
-       \\"nodeName\\": \\"html\\",
-       \\"raw\\": \\"\\",
-       \\"startCol\\": 103,
-       \\"startLine\\": 1,
-       \\"startOffset\\": 102,
-       \\"type\\": \\"omittedtag\\",
-     },
-     Object {
-       \\"endCol\\": 103,
-       \\"endLine\\": 1,
-       \\"endOffset\\": 102,
-       \\"isFragment\\": false,
-       \\"isGhost\\": true,
-       \\"namespace\\": \\"http://www.w3.org/1999/xhtml\\",
-       \\"nodeName\\": \\"head\\",
-       \\"raw\\": \\"\\",
-       \\"startCol\\": 103,
-       \\"startLine\\": 1,
-       \\"startOffset\\": 102,
-       \\"type\\": \\"omittedtag\\",
-     },
-     Object {
-       \\"endCol\\": 103,
-       \\"endLine\\": 1,
-       \\"endOffset\\": 102,
-       \\"isFragment\\": false,
-       \\"isGhost\\": true,
-       \\"namespace\\": \\"http://www.w3.org/1999/xhtml\\",
-       \\"nodeName\\": \\"body\\",
-       \\"raw\\": \\"\\",
-       \\"startCol\\": 103,
-       \\"startLine\\": 1,
-       \\"startOffset\\": 102,
-       \\"type\\": \\"omittedtag\\",
-     },
    ],
  }"
`;

exports[`parser html parser difference 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`parser invalid tag 1`] = `
Object {
  "isFragment": true,
  "nodeList": Array [
    Object {
      "attributes": Array [],
      "endCol": 5,
      "endLine": 1,
      "endOffset": 4,
      "endSpace": Object {
        "endCol": 4,
        "endLine": 1,
        "endOffset": 3,
        "raw": "",
        "startCol": 4,
        "startLine": 1,
        "startOffset": 3,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "a",
      "raw": "<a/>",
      "selfClosingSolidus": Object {
        "endCol": 4,
        "endLine": 1,
        "endOffset": 3,
        "raw": "/",
        "startCol": 3,
        "startLine": 1,
        "startOffset": 2,
      },
      "startCol": 1,
      "startLine": 1,
      "startOffset": 0,
      "tagCloseChar": "/>",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 8,
      "endLine": 1,
      "endOffset": 7,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "</>",
      "startCol": 5,
      "startLine": 1,
      "startOffset": 4,
      "type": "text",
    },
  ],
  "parseError": "Only void and foreign elements can be self closed \\"a\\" (\\"[ERROR ->]<a/></>\\"): angular-html-parser@0:0
Unexpected closing tag \\"\\". It may happen when the tag has already been closed by another tag. For more info see https://www.w3.org/TR/html5/syntax.html#closing-elements-that-have-implied-end-tags (\\"<a/>[ERROR ->]</>\\"): angular-html-parser@0:4",
}
`;

exports[`parser parse 1`] = `
Object {
  "isFragment": false,
  "nodeList": Array [
    Object {
      "endCol": 16,
      "endLine": 1,
      "endOffset": 15,
      "isFragment": false,
      "isGhost": false,
      "name": "html",
      "nodeName": "#doctype",
      "publicId": "",
      "raw": "<!DOCTYPE html>",
      "startCol": 1,
      "startLine": 1,
      "startOffset": 0,
      "systemId": "",
      "type": "doctype",
    },
    Object {
      "endCol": 9,
      "endLine": 2,
      "endOffset": 24,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
        ",
      "startCol": 16,
      "startLine": 1,
      "startOffset": 15,
      "type": "text",
    },
    Object {
      "endCol": 72,
      "endLine": 6,
      "endOffset": 403,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#comment",
      "raw": "<![CDATA[ Within this Character Data block I can use double dashes as
        much as I want (along with <, &, ', and \\") *and* %MyParamEntity; will be
        expanded to the text \\"Has been expanded\\" ... however, I can't use the
        CEND sequence. If I need to use CEND I must escape one of the brackets
        or the greater-than sign using concatenated CDATA sections. ]]>",
      "startCol": 9,
      "startLine": 2,
      "startOffset": 24,
      "type": "comment",
    },
    Object {
      "endCol": 9,
      "endLine": 7,
      "endOffset": 412,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
        ",
      "startCol": 72,
      "startLine": 6,
      "startOffset": 403,
      "type": "text",
    },
    Object {
      "endCol": 35,
      "endLine": 7,
      "endOffset": 438,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#comment",
      "raw": "<!-- top level comment -->",
      "startCol": 9,
      "startLine": 7,
      "startOffset": 412,
      "type": "comment",
    },
    Object {
      "endCol": 9,
      "endLine": 8,
      "endOffset": 447,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
        ",
      "startCol": 35,
      "startLine": 7,
      "startOffset": 438,
      "type": "text",
    },
    Object {
      "attributes": Array [
        Object {
          "endCol": 23,
          "endLine": 8,
          "endOffset": 462,
          "endQuote": Object {
            "endCol": 23,
            "endLine": 8,
            "endOffset": 462,
            "raw": "\\"",
            "startCol": 22,
            "startLine": 8,
            "startOffset": 461,
          },
          "equal": Object {
            "endCol": 19,
            "endLine": 8,
            "endOffset": 458,
            "raw": "=",
            "startCol": 18,
            "startLine": 8,
            "startOffset": 457,
          },
          "isInvalid": false,
          "name": Object {
            "endCol": 18,
            "endLine": 8,
            "endOffset": 457,
            "raw": "lang",
            "startCol": 14,
            "startLine": 8,
            "startOffset": 453,
          },
          "raw": "lang=\\"en\\"",
          "spacesAfterEqual": Object {
            "endCol": 19,
            "endLine": 8,
            "endOffset": 458,
            "raw": "",
            "startCol": 19,
            "startLine": 8,
            "startOffset": 458,
          },
          "spacesBeforeEqual": Object {
            "endCol": 18,
            "endLine": 8,
            "endOffset": 457,
            "raw": "",
            "startCol": 18,
            "startLine": 8,
            "startOffset": 457,
          },
          "spacesBeforeName": Object {
            "endCol": 14,
            "endLine": 8,
            "endOffset": 453,
            "raw": "",
            "startCol": 14,
            "startLine": 8,
            "startOffset": 453,
          },
          "startCol": 14,
          "startLine": 8,
          "startOffset": 453,
          "startQuote": Object {
            "endCol": 20,
            "endLine": 8,
            "endOffset": 459,
            "raw": "\\"",
            "startCol": 19,
            "startLine": 8,
            "startOffset": 458,
          },
          "type": "html-attr",
          "value": Object {
            "endCol": 22,
            "endLine": 8,
            "endOffset": 461,
            "raw": "en",
            "startCol": 20,
            "startLine": 8,
            "startOffset": 459,
          },
        },
      ],
      "endCol": 25,
      "endLine": 8,
      "endOffset": 463,
      "endSpace": Object {
        "endCol": 24,
        "endLine": 8,
        "endOffset": 462,
        "raw": "",
        "startCol": 24,
        "startLine": 8,
        "startOffset": 462,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "html",
      "raw": "<html lang=\\"en\\">",
      "selfClosingSolidus": Object {
        "endCol": 24,
        "endLine": 8,
        "endOffset": 462,
        "raw": "",
        "startCol": 24,
        "startLine": 8,
        "startOffset": 462,
      },
      "startCol": 9,
      "startLine": 8,
      "startOffset": 447,
      "tagCloseChar": ">",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 11,
      "endLine": 9,
      "endOffset": 474,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
          ",
      "startCol": 25,
      "startLine": 8,
      "startOffset": 463,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 17,
      "endLine": 9,
      "endOffset": 480,
      "endSpace": Object {
        "endCol": 16,
        "endLine": 9,
        "endOffset": 479,
        "raw": "",
        "startCol": 16,
        "startLine": 9,
        "startOffset": 479,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "body",
      "raw": "<body>",
      "selfClosingSolidus": Object {
        "endCol": 16,
        "endLine": 9,
        "endOffset": 479,
        "raw": "",
        "startCol": 16,
        "startLine": 9,
        "startOffset": 479,
      },
      "startCol": 11,
      "startLine": 9,
      "startOffset": 474,
      "tagCloseChar": ">",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 13,
      "endLine": 10,
      "endOffset": 493,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
            ",
      "startCol": 17,
      "startLine": 9,
      "startOffset": 480,
      "type": "text",
    },
    Object {
      "endCol": 34,
      "endLine": 10,
      "endOffset": 514,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#comment",
      "raw": "<!-- body comment -->",
      "startCol": 13,
      "startLine": 10,
      "startOffset": 493,
      "type": "comment",
    },
    Object {
      "endCol": 13,
      "endLine": 11,
      "endOffset": 527,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
            ",
      "startCol": 34,
      "startLine": 10,
      "startOffset": 514,
      "type": "text",
    },
    Object {
      "attributes": Array [
        Object {
          "endCol": 35,
          "endLine": 11,
          "endOffset": 549,
          "isDuplicatable": false,
          "isFragment": false,
          "isGhost": false,
          "potentialName": "type",
          "potentialValue": "text",
          "raw": "[type]=\\"'text'\\"",
          "startCol": 20,
          "startLine": 11,
          "startOffset": 534,
          "type": "ps-attr",
          "valueType": "string",
        },
      ],
      "endCol": 38,
      "endLine": 11,
      "endOffset": 552,
      "endSpace": Object {
        "endCol": 37,
        "endLine": 11,
        "endOffset": 551,
        "raw": "",
        "startCol": 37,
        "startLine": 11,
        "startOffset": 551,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "input",
      "raw": "<input [type]=\\"'text'\\" />",
      "selfClosingSolidus": Object {
        "endCol": 37,
        "endLine": 11,
        "endOffset": 551,
        "raw": " /",
        "startCol": 35,
        "startLine": 11,
        "startOffset": 549,
      },
      "startCol": 13,
      "startLine": 11,
      "startOffset": 527,
      "tagCloseChar": "/>",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 13,
      "endLine": 12,
      "endOffset": 565,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
            ",
      "startCol": 38,
      "startLine": 11,
      "startOffset": 552,
      "type": "text",
    },
    Object {
      "attributes": Array [
        Object {
          "endCol": 31,
          "endLine": 12,
          "endOffset": 584,
          "endQuote": Object {
            "endCol": 31,
            "endLine": 12,
            "endOffset": 584,
            "raw": "\\"",
            "startCol": 30,
            "startLine": 12,
            "startOffset": 583,
          },
          "equal": Object {
            "endCol": 24,
            "endLine": 12,
            "endOffset": 577,
            "raw": "=",
            "startCol": 23,
            "startLine": 12,
            "startOffset": 576,
          },
          "isInvalid": false,
          "name": Object {
            "endCol": 23,
            "endLine": 12,
            "endOffset": 576,
            "raw": "type",
            "startCol": 19,
            "startLine": 12,
            "startOffset": 572,
          },
          "raw": "type=\\"radio\\"",
          "spacesAfterEqual": Object {
            "endCol": 24,
            "endLine": 12,
            "endOffset": 577,
            "raw": "",
            "startCol": 24,
            "startLine": 12,
            "startOffset": 577,
          },
          "spacesBeforeEqual": Object {
            "endCol": 23,
            "endLine": 12,
            "endOffset": 576,
            "raw": "",
            "startCol": 23,
            "startLine": 12,
            "startOffset": 576,
          },
          "spacesBeforeName": Object {
            "endCol": 19,
            "endLine": 12,
            "endOffset": 572,
            "raw": "",
            "startCol": 19,
            "startLine": 12,
            "startOffset": 572,
          },
          "startCol": 19,
          "startLine": 12,
          "startOffset": 572,
          "startQuote": Object {
            "endCol": 25,
            "endLine": 12,
            "endOffset": 578,
            "raw": "\\"",
            "startCol": 24,
            "startLine": 12,
            "startOffset": 577,
          },
          "type": "html-attr",
          "value": Object {
            "endCol": 30,
            "endLine": 12,
            "endOffset": 583,
            "raw": "radio",
            "startCol": 25,
            "startLine": 12,
            "startOffset": 578,
          },
        },
        Object {
          "endCol": 49,
          "endLine": 12,
          "endOffset": 601,
          "isDuplicatable": false,
          "isFragment": false,
          "isGhost": false,
          "potentialName": "checked",
          "potentialValue": "true",
          "raw": "[checked]=\\"true\\"",
          "startCol": 33,
          "startLine": 12,
          "startOffset": 585,
          "type": "ps-attr",
          "valueType": "boolean",
        },
      ],
      "endCol": 52,
      "endLine": 12,
      "endOffset": 604,
      "endSpace": Object {
        "endCol": 51,
        "endLine": 12,
        "endOffset": 603,
        "raw": "",
        "startCol": 51,
        "startLine": 12,
        "startOffset": 603,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "input",
      "raw": "<input type=\\"radio\\" [checked]=\\"true\\" />",
      "selfClosingSolidus": Object {
        "endCol": 51,
        "endLine": 12,
        "endOffset": 603,
        "raw": " /",
        "startCol": 49,
        "startLine": 12,
        "startOffset": 601,
      },
      "startCol": 13,
      "startLine": 12,
      "startOffset": 565,
      "tagCloseChar": "/>",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 13,
      "endLine": 13,
      "endOffset": 617,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
            ",
      "startCol": 52,
      "startLine": 12,
      "startOffset": 604,
      "type": "text",
    },
    Object {
      "attributes": Array [
        Object {
          "endCol": 32,
          "endLine": 13,
          "endOffset": 637,
          "endQuote": Object {
            "endCol": 32,
            "endLine": 13,
            "endOffset": 637,
            "raw": "\\"",
            "startCol": 31,
            "startLine": 13,
            "startOffset": 636,
          },
          "equal": Object {
            "endCol": 24,
            "endLine": 13,
            "endOffset": 629,
            "raw": "=",
            "startCol": 23,
            "startLine": 13,
            "startOffset": 628,
          },
          "isInvalid": false,
          "name": Object {
            "endCol": 23,
            "endLine": 13,
            "endOffset": 628,
            "raw": "type",
            "startCol": 19,
            "startLine": 13,
            "startOffset": 624,
          },
          "raw": "type=\\"number\\"",
          "spacesAfterEqual": Object {
            "endCol": 24,
            "endLine": 13,
            "endOffset": 629,
            "raw": "",
            "startCol": 24,
            "startLine": 13,
            "startOffset": 629,
          },
          "spacesBeforeEqual": Object {
            "endCol": 23,
            "endLine": 13,
            "endOffset": 628,
            "raw": "",
            "startCol": 23,
            "startLine": 13,
            "startOffset": 628,
          },
          "spacesBeforeName": Object {
            "endCol": 19,
            "endLine": 13,
            "endOffset": 624,
            "raw": "",
            "startCol": 19,
            "startLine": 13,
            "startOffset": 624,
          },
          "startCol": 19,
          "startLine": 13,
          "startOffset": 624,
          "startQuote": Object {
            "endCol": 25,
            "endLine": 13,
            "endOffset": 630,
            "raw": "\\"",
            "startCol": 24,
            "startLine": 13,
            "startOffset": 629,
          },
          "type": "html-attr",
          "value": Object {
            "endCol": 31,
            "endLine": 13,
            "endOffset": 636,
            "raw": "number",
            "startCol": 25,
            "startLine": 13,
            "startOffset": 630,
          },
        },
        Object {
          "endCol": 47,
          "endLine": 13,
          "endOffset": 651,
          "isDuplicatable": false,
          "isFragment": false,
          "isGhost": false,
          "potentialName": "value",
          "potentialValue": "123",
          "raw": "[value]=\\"123\\"",
          "startCol": 34,
          "startLine": 13,
          "startOffset": 638,
          "type": "ps-attr",
          "valueType": "number",
        },
      ],
      "endCol": 50,
      "endLine": 13,
      "endOffset": 654,
      "endSpace": Object {
        "endCol": 49,
        "endLine": 13,
        "endOffset": 653,
        "raw": "",
        "startCol": 49,
        "startLine": 13,
        "startOffset": 653,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "input",
      "raw": "<input type=\\"number\\" [value]=\\"123\\" />",
      "selfClosingSolidus": Object {
        "endCol": 49,
        "endLine": 13,
        "endOffset": 653,
        "raw": " /",
        "startCol": 47,
        "startLine": 13,
        "startOffset": 651,
      },
      "startCol": 13,
      "startLine": 13,
      "startOffset": 617,
      "tagCloseChar": "/>",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 13,
      "endLine": 14,
      "endOffset": 667,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
            ",
      "startCol": 50,
      "startLine": 13,
      "startOffset": 654,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 18,
      "endLine": 14,
      "endOffset": 672,
      "endSpace": Object {
        "endCol": 17,
        "endLine": 14,
        "endOffset": 671,
        "raw": "",
        "startCol": 17,
        "startLine": 14,
        "startOffset": 671,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "div",
      "raw": "<div>",
      "selfClosingSolidus": Object {
        "endCol": 17,
        "endLine": 14,
        "endOffset": 671,
        "raw": "",
        "startCol": 17,
        "startLine": 14,
        "startOffset": 671,
      },
      "startCol": 13,
      "startLine": 14,
      "startOffset": 667,
      "tagCloseChar": ">",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 15,
      "endLine": 16,
      "endOffset": 707,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
              Text1
              ",
      "startCol": 18,
      "startLine": 14,
      "startOffset": 672,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 21,
      "endLine": 16,
      "endOffset": 713,
      "endSpace": Object {
        "endCol": 20,
        "endLine": 16,
        "endOffset": 712,
        "raw": "",
        "startCol": 20,
        "startLine": 16,
        "startOffset": 712,
      },
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "span",
      "raw": "<span>",
      "selfClosingSolidus": Object {
        "endCol": 20,
        "endLine": 16,
        "endOffset": 712,
        "raw": "",
        "startCol": 20,
        "startLine": 16,
        "startOffset": 712,
      },
      "startCol": 15,
      "startLine": 16,
      "startOffset": 707,
      "tagCloseChar": ">",
      "tagOpenChar": "<",
      "type": "starttag",
    },
    Object {
      "endCol": 25,
      "endLine": 16,
      "endOffset": 717,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "Text",
      "startCol": 21,
      "startLine": 16,
      "startOffset": 713,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 32,
      "endLine": 16,
      "endOffset": 724,
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "span",
      "raw": "</span>",
      "startCol": 25,
      "startLine": 16,
      "startOffset": 717,
      "tagCloseChar": ">",
      "tagOpenChar": "</",
      "type": "endtag",
    },
    Object {
      "endCol": 13,
      "endLine": 18,
      "endOffset": 757,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
              Text2
            ",
      "startCol": 32,
      "startLine": 16,
      "startOffset": 724,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 19,
      "endLine": 18,
      "endOffset": 763,
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "div",
      "raw": "</div>",
      "startCol": 13,
      "startLine": 18,
      "startOffset": 757,
      "tagCloseChar": ">",
      "tagOpenChar": "</",
      "type": "endtag",
    },
    Object {
      "endCol": 11,
      "endLine": 19,
      "endOffset": 774,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
          ",
      "startCol": 19,
      "startLine": 18,
      "startOffset": 763,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 18,
      "endLine": 19,
      "endOffset": 781,
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "body",
      "raw": "</body>",
      "startCol": 11,
      "startLine": 19,
      "startOffset": 774,
      "tagCloseChar": ">",
      "tagOpenChar": "</",
      "type": "endtag",
    },
    Object {
      "endCol": 9,
      "endLine": 20,
      "endOffset": 790,
      "isFragment": false,
      "isGhost": false,
      "nodeName": "#text",
      "raw": "
        ",
      "startCol": 18,
      "startLine": 19,
      "startOffset": 781,
      "type": "text",
    },
    Object {
      "attributes": Array [],
      "endCol": 16,
      "endLine": 20,
      "endOffset": 797,
      "isFragment": false,
      "isGhost": false,
      "namespace": "http://www.w3.org/1999/xhtml",
      "nodeName": "html",
      "raw": "</html>",
      "startCol": 9,
      "startLine": 20,
      "startOffset": 790,
      "tagCloseChar": ">",
      "tagOpenChar": "</",
      "type": "endtag",
    },
  ],
}
`;
